### PR TITLE
Test sending messages between friends

### PR DESCRIPTION
**Pull request by rohanjr**

### CODE REVIEW

File: ui/src/index.test.tsx [hunk 1]
```diff
@@ -244,3 +272,41 @@ test('error when adding existing friend', async () => {
...
+
+  const page1 = await newUiPage();
+  await login(page1, party1);
+
+  const page2 = await newUiPage();
+  await login(page2, party2);
+
+  // Once Party 2 is a friend of Party 1, Party 2 can send Party 1 a message.
+  await addFriend(page1, party2);
+  await sendMessageToFirst(page2, `Hey ${party1}!`);
```
Comment (cocreature) [inline ui/src/index.test.tsx]: Might be nice to add a negative test case where you try to send a message before the `addFriend` call. Doesn’t have to be in this PR.

File: ui/src/index.test.tsx [hunk 2]
```diff
@@ -129,6 +129,34 @@ const addFriend = async (page: Page, friendName: string) => {
...
+  await page.type('.test-select-message-content', content);
+
+  // Click send and wait for the request to complete.
+  await page.click('.test-select-message-send-button');
+  await page.waitForSelector('.test-select-message-send-button:not(.loading)');
+}
+
+// Assuming there is *at least one* message on the page,
+// count the number of messages.
+// The restriction against zero messages is because we need to wait on the
```
Comment (cocreature) [inline ui/src/index.test.tsx]: I’m not sure this is reliable either sadly? It seems like this only works if all messages appear at exactly the same time.
If you do something like first add one message. Then add another one and then call `countMessageNotZero` the `waitForSelector` check will immediately return due to the first message but potentially before you received the second message.

Assuming that is actually a problem we should at least add a note for now. Could we make this reliable and also support the `0` case by waiting until a loading indicator has disappeared or is there nothing that indicates that a request is in progress atm?

Comment (rohanjr) [reply ui/src/index.test.tsx]: There is a loading indicator on the message Send button, and we wait for it to disappear at the end of `sendMessageToFirst`.

> If you do something like first add one message. Then add another one and then call countMessageNotZero the waitForSelector check will immediately return due to the first message but potentially before you received the second message.

This makes sense but after several runs I don't actually see this problem. For some reason the
`await page.waitForSelector('.test-select-message');`
line in `countMessagesNotZero` seems to do the trick in waiting for new messages, whereas
`await page.waitForSelector('.test-select-message-send-button:not(.loading)');`
(which is at the end of `sendMessageToFirst`) does not. I tried replacing the first line with the second in `countMessages`: that makes the 0 case work but the 2 case flaky.

Sorry I don't have a better answer to this at the moment. I will put some more information in the comment for now.

Comment (rohanjr) [reply ui/src/index.test.tsx]: Ok, I might have a better solution to this. Putting a CSS class in the message list (as well as the individual items) and an `await page.waitForSelector('.test-select-message-list');` seems to be effective. I'll test a little more and push another commit.

Comment (rohanjr) [reply ui/src/index.test.tsx]: Ah no, scratch that :( It seems that the message list is loaded before the items, but the items are loaded at the same time (just from what I'm observing).

File: ui/src/components/MessageList.tsx [hunk 3]
```diff
@@ -14,7 +14,9 @@ const MessageList: React.FC = () => {
       {messagesResult.contracts.map(message => {
         const {sender, receiver, content} = message.payload;
         return (
-          <ListItem key={message.contractId}>
+          <ListItem
+            className='test-select-message'
```
[the remark this answers is not in the document]
Comment (rohanjr) [reply ui/src/components/MessageList.tsx]: Fixed in the last commit

File: ui/src/index.test.tsx (hunk 2 above)
Comment (hurryabit) [reply ui/src/index.test.tsx]: Do we have a test to check that a second (or third message arrive)?

File: ui/src/index.test.tsx [hunk 4]
```diff
@@ -129,6 +129,35 @@ const addFriend = async (page: Page, friendName: string) => {
...
 }
 
+// Send a message to the first friend appearing in the dropdown menu.
+const sendMessageToFirst = async (page: Page, content: string) => {
+  // Select the first option in the dropdown menu.
+  const dropdown = '.test-select-message-receiver > .dropdown';
+  const item = `${dropdown} > .menu > .item`;
+  await page.click(dropdown);
+  await page.waitForSelector(item);
+  await page.click(item);
```
[the remark this answers is not in the document]
Comment (rohanjr) [reply ui/src/index.test.tsx]: It will take the first one. I didn't generalise this as it wasn't needed yet

File: ui/src/index.test.tsx (hunk 2 above)
Comment (rohanjr) [reply ui/src/index.test.tsx]: No, but I can add it

**Pull request merged**